contracts/BatchAuction.sol

```diff
@@ -59,24 +65,45 @@ contract BatchAuction is Ownable {
 
         address tokenAddress = tokenIdToAddressMap[tokenIndex];
         require(tokenAddress != address(0), "Requested token is not registered");
-
         require(
             ERC20(tokenAddress).transferFrom(msg.sender, address(this), amount), 
             "Unsuccessful transfer"
         );
-        
-        if (depositIndex != block.number / 20) {
-            depositIndex = block.number / 20;
-            emit IncrementDepositIndex(depositIndex);
-        }
 
+        uint depositSlot = this.depositIndex();
         uint16 accountId = publicKeyToAccountMap[msg.sender];
         bytes32 nextDepositHash = sha256(
-            abi.encodePacked(depositHashes[depositIndex].shaHash, accountId, tokenIndex, amount)
+            abi.encodePacked(depositHashes[depositSlot].shaHash, accountId, tokenIndex, amount)
         );
 
-        depositHashes[depositIndex] = DepositState({shaHash: nextDepositHash, applied: false});
+        depositHashes[depositSlot] = DepositState({shaHash: nextDepositHash, applied: false});
+        emit Deposit(accountId, tokenIndex, amount, depositSlot);
+    }
+
+    function depositIndex() public view returns (uint) {
+        return block.number / 20;
+    }
+
+    function stateIndex() public view returns (uint) {
+        return stateRoots.length - 1;
+    }
 
-        emit Deposit(accountId, tokenIndex, amount, depositIndex);
+    function applyDeposits(
+        uint slot,
+        bytes32 _currDepositHash,
+        bytes32 _currStateRoot,
+        bytes32 _newStateRoot
+    )
+        public onlyOwner()
+    {   
+        require(slot < this.depositIndex(), "Deposit slot must exist and be inactive");
```

Comment:
Is this needed, or are we secure by just requiring that depositHashes[slot].sha !=0

Reply:
This is needed for sure. We can't apply deposits before we know for sure that the deposit hash at this slot can never change again.